thavalon-server/src/game/snapshot.rs

```diff
@@ -0,0 +1,311 @@
+    pub role: Role,
+    pub info: RoleDetails,
+}
+
+/// Details about a mission. If the mission is in process, data may be missing.
+#[derive(Debug, Clone)]
+pub struct Mission {
+    /// Submitted proposals. Every proposal that has been voted on will have a corresponding entry in `voting_results`.
+    pub proposals: Vec<Proposal>,
+    /// Results from voting on proposals.
+    pub voting_results: Vec<VotingResults>,
+    /// If a proposal has been voted on and sent, this will be the index into `proposals` of the one that was sent. In most
+    /// cases, this will be the *last* proposal. However, on mission one it could be either of the two proposals.
+    pub sent_proposal: Option<usize>,
+    /// If the mission has gone, this has the results, including what cards were played and whether or not it passed. If Agravaine declares,
+    /// the mission's `passed` state may change again.
+    pub results: Option<MissionResults>,
+}
+
+/// The outcome of a mission, including details on which cards were played and whether or not the mission passed.
+#[derive(Debug, Clone)]
+pub struct MissionResults {
+    pub successes: usize,
+    pub fails: usize,
+    pub reverses: usize,
+    pub questing_beasts: usize,
+    pub passed: bool,
+    pub agravaine_declared: bool,
+}
+
+#[derive(Debug, Clone)]
+pub struct Proposal {
+    pub proposed_by: String,
+    pub players: HashSet<String>,
+}
+
+/// Results of voting on a mission. Normally, this includes exactly who upvoted or downvoted. If Maeve obscured the mission,
+/// only counts are known.
+#[derive(Debug, Clone)]
+pub enum VotingResults {
+    Public {
+        upvotes: HashSet<String>,
+        downvotes: HashSet<String>,
+    },
+    Obscured {
+        upvotes: u32,
+        downvotes: u32,
+    },
+}
+
+impl GameSnapshot {
+    pub fn new() -> GameSnapshot {
+        GameSnapshot {
+            role_info: None,
+            missions: Vec::new(),
+            log: Vec::new(),
+        }
+    }
+
+    /// Looks up mission information by number. If the game has not reached `mission` yet, returns `None`.
+    pub fn mission(&self, mission: MissionNumber) -> Option<&Mission> {
+        self.missions.get((mission - 1) as usize)
+    }
+
+    /// Like [`Self::mission`], but returns a mutable reference
+    fn mission_mut(&mut self, mission: MissionNumber) -> &mut Mission {
+        &mut self.missions[(mission - 1) as usize]
+    }
+
+    /// The number of the mission currently in progress
+    pub fn current_mission(&self) -> MissionNumber {
+        self.missions.len() as MissionNumber
+    }
+
+    /// Get a mutable reference to the current mission.alloc
+    /// 
+    /// # Panics
+    /// If there is *no* current mission, which would only happen if messages were received in an invalid
+    /// order.
+    fn current_mut(&mut self) -> &mut Mission {
+        self.missions.last_mut().unwrap()
+    }
+
+    pub fn on_message(&mut self, message: Message) -> Result<(), SnapshotError> {
```

Comment:
This function is public but doesn't have a header. Should we add one here?

Reply:
Yeah, will do